### ISSUE TITLE
Log command frequency to improve keymap

### ISSUE THREAD

**Issue by eugenesvk** [helix-editor/helix#1503]
I'd like to improve Helix keymap by shifting more frequently used commands to more convenient keys. I've already partially done it based on the [Emacs command stats](http://xahlee.info/emacs/emacs/command-frequency.html) collected by Xah Lee, but it obviously doesn't cover every command in Helix.
Would it be possible to add a logging feature that would create a list of every single executed command within Helix that could then be used to calculate command frequencies and make data-based keymap design decisions?

**Comment by eduardorittner** [comment]
This seems trivial to implement by just doing a `log::info` every time a key event is matched to a command. All the executed commands would then be available in the default helix log file, and to pull them out you could use a simple python script or something. What would be nicer in my opinion would be for helix to have built-in support for this, maybe saving the stats on a separate file only for them and being able to show all-time stats or stats for the current. Thoughts?

**Comment by eugenesvk** [comment]
yeah, the built-in data collection would hopefully be more than just logging, I thought of a simple database
For each unique user keymap you'd have file with a copy of said keymap and the following data
- for each mode
  - for each keybind
    - for each unique list of commands invoked by this keybind
      - for each date (with a daily/weekly? resolution? not sure how much precision matters here)
        - for each syntax
          - data: # of times invoked

And you'd build that structure in memory and merge/flush to disk form time to time not to save on each keystroke.

Then you'd have a good overview of how you use your keymap.

Maybe you could also add a 2nd item outside of insert mode (to avoid being able to recreate what you type): incremental id so that you could analyze sequences? But haven't thought much beyond that, for the original purpose this should be enough.

For other purposes you might need more (e.g., commands you invoke via`:` etc)

**Comment by eduardorittner** [comment]
I'm not sure I understand how the same keybind in the same mode can invoke different commands, but that's probably because I'm not really familiar with how helix handles them. Also, what would different syntaxes be for the same command, keybind and mode? Also, this would only store static commands, right? Not `TypableCommands`.

**Comment by eugenesvk** [comment]
You could have a menu so `z` invokes a menu and the next `z` selects a command, so same key - different commands within the same mode.
Syntax I meant the active treesitter grammar in case you wanted to see, e.g., the difference between navigating plain text and Rust.

**Comment by eduardorittner** [comment]
Makes sense. I'm not really sure how to handle the different menus, since they don't seem to have the `KeyTrie` commands, instead they handle key events with a match expression, and the actions they take don't map cleanly to `MappableCommands`. So I see two possibilities: Either ignore these menus and just log the `MappableCommands`, which would be simpler but not as complete. Or find a way to log these keyevents inside each `Component` impl. I don't think I'd know how to implement the second case without littering every match arm from the Components with logs.

**Comment by eugenesvk** [comment]
Maybe the functions in the right arms of the matches could also accept some kind of "context" argument that would include the `ch`ar that invoked it and the menu id?
So it the z-z example it would be : `{ch: 'z', parent_id:"menu_z"}`?

Then instead of adding "log" in every match you could add log to this single action function

**Comment by eduardorittner** [comment]
I'm still thinking about this, and have some more questions:
- Would the data look like some nested HashMaps, or a full-blown SQL database?
- Wouldn't this have a non-negligible impact on performance, and so preferably be opt-in? Depending on how this is handled, even when it's disabled it would still impact performance (a check on every keystroke).
- Is this worth it? I think recording the frequency of commands and binds would be pretty cool, but is it useful for enough users that the increase in complexity on keymap handling logic is justifiable? 
- Not a question, just adding that implementing this in a unified way for both keymaps and typable commands across different contexts (prompt, markdown ui, completion, etc) is really complex due to how mode keymaps, context-dependent keymaps and typable commands are all handled in very dinstinct ways.

**Comment by eduardorittner** [comment]
I really like the data stream idea, because on every keystroke you'd have at most 2 or 3 vec pushes, whereas with more structured data you might have several hash map gets and/or inserts. Disk saves could be batched every so often like you said.
In my head it's easier to save the individual events (enter/exit contenxt, key-pressed, etc.) and only produce the structured data on-demand, with a command like ":key-stats" or something. But I can also see the value of producing it at runtime and storing only the serialized data. Either way I think I'll try and work on producing these events first and come back to how to store them later.

**Comment by eduardorittner** [comment]
Also on your other answers, yeah I was thinking a boolean check on every keystroke would be the simplest and also allow for toggling at runtime. And ignoring typable commands sounds good.

**Comment by eugenesvk** [comment]
The only challenge with a stream could be missing some state change and thus making all subsequent events impossible to decode (theoretical example: a mode change happened as a result of some typable command but that mode change didn't emit any stream event since it was a separate code path from the mode change that is triggered by a keypress)?

Maybe could inject "current state is ..." for sync and help discard data with an error message where tracking state from stream doesn't match the resulting synchronization "current state"
